File: server/spotify-commands.js
/**
 * NFG Crash Spotify chat commands — prefixed so they do not clash with Hangman (!hsong / !hqueue).
 *
 * Crash:  !csong  !cqueue  !caddsong  (long: !crashsong !crashqueue !crashaddsong)
 * Hangman uses: !hsong !hqueue !haddsong (see hangman v2/server.py)
 *
 * Bare !song / !queue / !addsong are rejected with a hint (both games may be live).
 */

const CRASH_QUEUE_RE =
  /^!(?:c(?:song|queue|addsong)|crash(?:song|queue|addsong))\b(.*)$/i;

const LEGACY_GENERIC_RE = /^!(?:song|queue|addsong)\b/i;

function parseCrashSpotifyQueueCommand(text) {
  const raw = String(text || "").trim();
  if (!raw) return null;
  if (LEGACY_GENERIC_RE.test(raw)) {
    return {
      rejected: true,
      reason: "ambiguous",
      help: "Use !csong or !cqueue for NFG Crash (Hangman uses !hsong / !hqueue).",
    };
  }
  const m = raw.match(CRASH_QUEUE_RE);
  if (!m) return null;
  const full = raw.slice(1).toLowerCase();
  let command = "song";
  if (full.startsWith("cqueue") || full.startsWith("crashqueue")) command = "queue";
  else if (full.startsWith("caddsong") || full.startsWith("crashaddsong")) command = "addsong";
  else if (full.startsWith("csong") || full.startsWith("crashsong")) command = "song";
  return {
    command,
    query: String(m[1] || "").trim(),
  };
}

module.exports = { parseCrashSpotifyQueueCommand };

File: hangman v2/spotify_commands.py
"""Hangman Spotify chat command parsing — prefixed to avoid clash with NFG Crash."""
from __future__ import annotations

import re
from typing import Optional

# !hsong !hqueue !haddsong  or  !hangmansong !hangmanqueue !hangmanaddsong
_HANGMAN_QUEUE_RE = re.compile(
    r"^!(?:h(?:song|queue|addsong)|hangman(?:song|queue|addsong))\b(.*)$",
    re.I,
)
_LEGACY_GENERIC_RE = re.compile(r"^!(?:song|queue|addsong)\b", re.I)

# Admin: !hqueueallow !hqueuedeny !hqueueremove !hqueuelist (+ long forms)
_HANGMAN_ADMIN_CMDS = frozenset(
    {
        "!hqueueallow",
        "!hqueuedeny",
        "!hqueueremove",
        "!hqueuelist",
        "!hangmanqueueallow",
        "!hangmanqueuedeny",
        "!hangmanqueueremove",
        "!hangmanqueuelist",
    }
)


def hangman_spotify_queue_match(text: str) -> Optional[re.Match[str]]:
    raw = (text or "").strip()
    if not raw:
        return None
    if _LEGACY_GENERIC_RE.match(raw):
        return None
    return _HANGMAN_QUEUE_RE.match(raw)


def hangman_spotify_command_name(text: str) -> str:
    """song | queue | addsong"""
    raw = (text or "").strip().lower()
    if raw.startswith("!hqueue") or raw.startswith("!hangmanqueue"):
        return "queue"
    if raw.startswith("!haddsong") or raw.startswith("!hangmanaddsong"):
        return "addsong"
    return "song"


def hangman_spotify_admin_command(text: str) -> Optional[str]:
    parts = (text or "").strip().split(maxsplit=1)
    if not parts:
        return None
    cmd = parts[0].lower()
    if cmd in _HANGMAN_ADMIN_CMDS:
        return cmd
    return None


def legacy_spotify_hint() -> str:
    return "[Spotify] Use !hsong or !hqueue for Hangman (NFG Crash uses !csong / !cqueue)."


def hangman_spotify_usage_hint() -> str:
    return (
        "[Spotify] Hangman: !hsong / !hqueue / !haddsong "
        "(or !hangmansong …). Broadcaster: !hqueueallow @viewer"
    )
